### PR TITLE
nixos/nebula: wait for start notification to prevent startup race

### PR DESCRIPTION
**Pull request by numinit**
## Description of changes

Support https://github.com/slackhq/nebula/pull/929. Marks the service as started when the tunnel is up.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by numinit** [comment]
cc @Jaculabilis

**Comment by numinit** [comment]
Enabled by #272581

File: nixos/modules/services/networking/nebula.nix [hunk 1]
```diff
@@ -196,7 +196,8 @@ in
             before = [ "sshd.service" ];
             wantedBy = [ "multi-user.target" ];
             serviceConfig = {
-              Type = "simple";
+              Type = "notify";
+              NotifyAccess = "main";
```
Comment (jian-lin) [inline nixos/modules/services/networking/nebula.nix:200]: `NotifyAccess = "main";` is implied if `Type = "notify";` is set according to `man systemd.service`.  What about removing this line?

Comment (numinit) [reply nixos/modules/services/networking/nebula.nix:200]: Sounds good. Did it because of the Nebula example code but am happy to remove it.

**Review by jian-lin — APPROVED** [review]
LGTM

Can https://github.com/NixOS/nixpkgs/pull/273273 be reverted after we have this change?

**Comment by numinit** [comment]
@jian-lin I think #273273 is probably good to keep because it's defensive. We should wait for the service to be up -- yes, it may be redundant, but I'm not sure if it's redundant with respect to `systemctl restart`.

**Comment by jian-lin** [comment]
> yes, it may be redundant, but I'm not sure if it's redundant with respect to systemctl restart

It is indeed redundant.  `systemctl restart` is `systemctl stop` followed by `systemctl start` according to `man systemctl`.

> I think https://github.com/NixOS/nixpkgs/pull/273273 is probably good to keep because it's defensive.

I see #273273 as a workaround for an issue in the nebula service and this PR fixes that issue properly.

**Review by NickCao — APPROVED** [review]

**Comment by numinit** [comment]
I created #274087. Once it's merged I'll do a manual backport PR with all these changes, including Nebula 1.8 which adds the more predictable systemd service start behavior.

**Pull request merged**